storage/invoke_test.go

```diff
@@ -121,8 +169,94 @@ func TestInvoke(t *testing.T) {
 				return test.finalErr
 			}
 			got := run(ctx, call, test.retry, test.isIdempotentValue)
-			if got != test.finalErr {
-				s.Errorf("%+v: got %v, want %v", test, got, test.finalErr)
+			if test.expectFinalErr && got != test.finalErr {
+				s.Errorf("got %v, want %v", got, test.finalErr)
+			} else if !test.expectFinalErr && got != test.initialErr {
+				s.Errorf("got %v, want %v", got, test.initialErr)
+			}
+		})
+	}
+}
+
+func TestShouldRetry(t *testing.T) {
+	t.Parallel()
+
+	for _, test := range []struct {
+		desc        string
+		inputErr    error
+		shouldRetry bool
+	}{
+		{
+			desc:        "googleapi.Error{Code: 0}",
+			inputErr:    &googleapi.Error{Code: 0},
+			shouldRetry: false,
+		},
+		{
+			desc:        "googleapi.Error{Code: 429}",
+			inputErr:    &googleapi.Error{Code: 429},
+			shouldRetry: true,
+		},
+		{
+			desc:        "errors.New(foo)",
+			inputErr:    errors.New("foo"),
+			shouldRetry: false,
+		},
+		{
+			desc:        "googleapi.Error{Code: 518}",
+			inputErr:    &googleapi.Error{Code: 518},
+			shouldRetry: true,
+		},
+		{
+			desc:        "googleapi.Error{Code: 599}",
+			inputErr:    &googleapi.Error{Code: 599},
+			shouldRetry: true,
+		},
+		{
+			desc:        "googleapi.Error{Code: 428}",
+			inputErr:    &googleapi.Error{Code: 428},
+			shouldRetry: false,
+		},
+		{
+			desc:        "googleapi.Error{Code: 518}",
+			inputErr:    &googleapi.Error{Code: 518},
+			shouldRetry: true,
+		},
+		{
+			desc:        "url.Error{Err: errors.New(\"connection refused\")}",
+			inputErr:    &url.Error{Op: "blah", URL: "blah", Err: errors.New("connection refused")},
+			shouldRetry: true,
+		},
+		{
+			desc:        "io.ErrUnexpectedEOF",
+			inputErr:    io.ErrUnexpectedEOF,
+			shouldRetry: true,
+		},
+		{
+			desc:        "Test unwrapping of a temporary error",
+			inputErr:    xerrors.Errorf("Test unwrapping of a temporary error: %w", &googleapi.Error{Code: 500}),
+			shouldRetry: true,
+		},
+		{
+			desc:        "Test unwrapping of a non-retriable error",
+			inputErr:    xerrors.Errorf("Test unwrapping of a non-retriable error: %w", &googleapi.Error{Code: 400}),
+			shouldRetry: false,
+		},
+		{
+			desc:        "googleapi.Error{Code: 400}",
+			inputErr:    &googleapi.Error{Code: 400},
+			shouldRetry: false,
+		},
+		{
+			desc:        "googleapi.Error{Code: 408}",
+			inputErr:    &googleapi.Error{Code: 408},
+			shouldRetry: false, // to be changed
+		},
```

Comment:
Could you add cases for retryable/non-retryable gRPC statuses?

Reply:
done! let me know if you think we should add more/all possible gRPC statuses

Reply:
looks good!